event_sourcery/event_store/backend.py

```diff
@@ -0,0 +1,174 @@
+        self, limit: int
+    ) -> Iterator[AbstractContextManager[RecordedRaw]]:
+        return iter([])
+
+
+T_co = TypeVar("T_co", covariant=True)
+T = TypeVar("T")
+
+
+class _Provider(Protocol[T_co]):
+    def __call__(self, container: "_Container") -> T_co: ...
+
+
+def singleton(provider: _Provider[T]) -> _Provider[T]:
+    result: T | None = None
+
+    @wraps(provider)
+    def _wrapper(container: "_Container") -> T:
+        nonlocal result
+        if result is not None:
+            return result
+        result = provider(container)
+        return result
+
+    return _wrapper
+
+
+def not_configured(error_message: str) -> _Provider[T]:
+    def _raise(container: "_Container") -> NoReturn:
+        raise NoProviderConfigured(error_message)  # pragma: no cover
+
+    return _raise
+
+
+class _Container(UserDict[type[T], _Provider[T] | T]):
+    def __getitem__(self, _type: type[T] | type) -> T:
+        provider = cast(_Provider[T], self.data[_type])
+        return provider(self)
+
+    def __setitem__(self, _type: type[T], value: T | _Provider[T]) -> None:
+        self.data[_type] = (
+            cast(_Provider[T], lambda _: value) if isinstance(value, _type) else value
+        )
+
+    def copy(self) -> Self:
+        new: Self = self.__class__()
+        new.data.update(self.data.copy())
+        return new
+
+
+class Backend(_Container):
+    tenant_id: TenantId
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.tenant_id = DEFAULT_TENANT
+        self[EventRegistry] = lambda _: EventRegistry()
+        self[EncryptionStrategy] = lambda _: NoEncryptionStrategy()
+        self[EncryptionKeyStorageStrategy] = (
+            lambda c: NoKeyStorageStrategy().scoped_for_tenant(c.tenant_id)
+        )
+        self[Encryption] = lambda c: Encryption(
+            registry=c[EventRegistry],
+            strategy=c[EncryptionStrategy],
+            key_storage=c[EncryptionKeyStorageStrategy],
+        )
+        self[Serde] = lambda c: Serde(
+            registry=c[EventRegistry],
+            encryption=c[Encryption],
+        )
+        self[StorageStrategy] = not_configured(
+            "Use one of pyES backends: SQLAlchemy, Django or KurrentDB",
+        )
+        self[EventStore] = lambda c: EventStore(
+            storage_strategy=c[StorageStrategy],
+            serde=c[Serde],
+        )
+        self[Outbox] = lambda c: Outbox(
+            strategy=c[OutboxStorageStrategy],
+            serde=c[Serde],
+        )
+        self[OutboxStorageStrategy] = lambda _: NoOutboxStorageStrategy()
+        self[SubscriptionStrategy] = not_configured(
+            "Use one of pyES backends: SQLAlchemy, Django or KurrentDB",
+        )
+        self[subscription.PositionPhase] = lambda c: subscription.SubscriptionBuilder(
+            c[Serde],
+            c[SubscriptionStrategy],
+        )
+
+    @property
+    def event_store(self) -> EventStore:
+        return cast(EventStore, self[EventStore])
```

Comment:
Why do you need casting?

Reply:
I think `_Container` annotation or even concept is wrong.
Right now mypy sees `self[]` as returning Any. I think we should talk about it.